lib/common/common/src/counter/counter_cell.rs

```diff
@@ -83,4 +83,73 @@ impl CounterCell {
     pub fn take(&self) -> usize {
         self.counter.take()
     }
+
+    /// Creates a write-back counter for best performance possible.
+    /// For more information on when and why to use, see [`WriteBackCounterCell`]
+    #[inline]
+    pub fn write_back_counter(&self) -> WritebackCounterCell {
+        WritebackCounterCell::new(self)
+    }
+}
+
+/// Performance optimized counter to measure hot-paths relieably. It accumulates it's current measurements
+/// inside a `usize` and writes the result back into the original counter on drop.
+///
+///
+/// ## Why and when should I use this instead of [`CounterCell`]?
+///
+/// The `CounterCell::incr_delta()` function is around twice as slow as counting a `usize` integer.
+/// This is because we have to copy the cells value, do the arithmetic and write the value back.
+/// Usually this is not a problem because it is still a very fast operation. In loops or hot-paths however
+/// it can become a considerable overhead we want to avoid as much as possible.
+///
+/// You should always prefer this over manually counting a loop because we might loose values when returning
+/// on an error, early-return or add such code in future and forget to adjust.
+///
+///
+/// ## When to *not* use this?
+///
+/// Because this writeback counter only writes its values into the original cell when dropped, this is
+/// not suitable if you directly need to read from a `CounterCell` within the same scope. This should
+/// however avoided as much as possible, because these structures are for collecting measurements and should
+/// be read from the initial [`HwMeasurementAcc`].
+pub struct WritebackCounterCell<'a> {
+    cell: &'a CounterCell,
+    counter: usize,
+}
+
+impl Drop for WritebackCounterCell<'_> {
+    #[inline]
+    fn drop(&mut self) {
+        self.cell.incr_delta(self.counter);
+    }
+}
+
+impl<'a> WritebackCounterCell<'a> {
+    #[inline]
+    fn new(cell: &'a CounterCell) -> Self {
+        Self { cell, counter: 0 }
+    }
+
+    #[inline]
+    pub fn incr_delta(&mut self, delta: usize) {
+        self.counter += delta;
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_write_back_counter() {
+        let cell = CounterCell::new();
+
+        {
+            let mut wb_counter = cell.write_back_counter();
+            wb_counter.incr_delta(4);
+        }
+
+        assert_eq!(cell.get(), 4)
```

Comment:
How do you feel about adding this assertion?

It's not necessary, but it would prove the type behaves like we tell it does.

```suggestion
        {
            let mut wb_counter = cell.write_back_counter();
            wb_counter.incr_delta(4);
            assert_eq!(cell.get(), 0);
        }

        assert_eq!(cell.get(), 4);
```